From 2da59647f3ba247468a547a0fd9bfcc257e51618 Mon Sep 17 00:00:00 2001
From: slackerlinux85 <slackerlinux85@gmail.com>
Date: Mon, 10 Nov 2014 13:11:38 +0000
Subject: [PATCH] MythThemedMenu: Add dependsexec option to the main menu

This option allows for menu entries that will only be visible if a particular
executable is found. For example the menu entry below will add an XBMC menu
entry if the XBMC executable is found at /usr/bin/xbmc.

<button>
  <type>TV</type>
  <text>XBMC</text>
  <description>Launch XBMC</description>
  <action>EXEC xbmc</action>
  <dependsexec>/usr/bin/xbmc</dependsexec>
</button>

Fixes #12021.

Signed-off-by: Paul Harrison <pharrison@mythtv.org>
---
 mythtv/libs/libmythui/myththemedmenu.cpp |   13 +++++++++++++
 mythtv/libs/libmythui/myththemedmenu.h   |    1 +
 2 files changed, 14 insertions(+)

diff --git a/mythtv/libs/libmythui/myththemedmenu.cpp b/mythtv/libs/libmythui/myththemedmenu.cpp
index 0d66514..269de11 100644
--- a/mythtv/libs/libmythui/myththemedmenu.cpp
+++ b/mythtv/libs/libmythui/myththemedmenu.cpp
@@ -7,6 +7,7 @@
 #include <QKeyEvent>
 #include <QDomDocument>
 #include <QFile>
+#include <QFileInfo>
 #include <QTextStream>
 
 // libmythui headers
@@ -509,6 +510,10 @@ void MythThemedMenu::parseThemeButton(QDomElement &element)
             {
                 addit = findDepends(getFirstText(info));
             }
+            else if (info.tagName() == "dependsexec")
+            {
+                addit = findDependsExec(getFirstText(info));
+            }
             else if (info.tagName() == "dependssetting")
             {
                 addit = GetMythDB()->GetNumSetting(getFirstText(info));
@@ -885,6 +890,14 @@ bool MythThemedMenu::findDepends(const QString &fileList)
     return false;
 }
 
+bool MythThemedMenu::findDependsExec(const QString &filename)
+{
+    QFileInfo filename_info(filename);
+    if (filename_info.exists() && filename_info.isFile() && filename_info.isExecutable())
+        return true;
+    return false;
+}
+
 /** \brief Queries the user for a password to enter a part of MythTV
  *         restricted by a password.
  *
diff --git a/mythtv/libs/libmythui/myththemedmenu.h b/mythtv/libs/libmythui/myththemedmenu.h
index e85b37a..831a8e8 100644
--- a/mythtv/libs/libmythui/myththemedmenu.h
+++ b/mythtv/libs/libmythui/myththemedmenu.h
@@ -103,6 +103,7 @@ class MUI_PUBLIC MythThemedMenu : public MythThemedMenuState
 
     bool handleAction(const QString &action, const QString &password = QString());
     bool findDepends(const QString &fileList);
+    bool findDependsExec(const QString &filename);
     QString findMenuFile(const QString &menuname);
 
     bool checkPinCode(const QString &password_setting);
-- 
1.7.10.2

